ocaml/xenopsd/xc/xenops_server_xen.ml

```diff
@@ -1041,6 +1057,15 @@ module HOST = struct
         }
     )
 
+  let stat () =
+    match !stat_cache with
+    | None ->
+        let s = stat' () in
```

Comment:
This may work, but might be problematic in some cases, e.g. what if a Xen or microcode live-patch changes the available features on the CPU and requires reevaluating the next-boot featureset?
In particular when features like TSX are removed they're still emulated (in a correct, but slow way by aborting it), and on next boot we'd want to remove the TSX cpuid bit completely so the guest doesn't even try to use TSX. This happens because the host has changed its featureset "live".

We'd probably need a way to invalidate this cache, e.g. toolstack restart might work, but might be useful to have something more targeted.

Reply:
XAPI should be aware when patch RPMs are applied through supported mechanisms, so it could send a command to xenopsd to invalidate the stat cache?

Reply:
The actions we support to some form or another are this:

- Livepatch Xen knowing about new functionality
- Late load ucode with new functionality
    - Raw features gets updated at this point, admin checks via `xen-cpuid`
- If everything is ok, admin elects to update the default/max policies via `xen-spec-ctrl` (name in need of change)
    - toolstack restart at this point to pick everything new up
- New VMs started pick up new features
    - Obviously, we'd like to include reboots here too so VMs have expected behaviour

The 3rd bullet point is very important.  It offers the admin a way out if something didn't happen as expected with the ucode load.

For the TSX case specifically, we do not hide the RTM/HLE CPUID bits during late loading.  The runtime complexity is just too great, as demonstrated by the number of bugs we've had around maintaining migration compatibility on a clean boot of Xen.

Reply:
You're right, but currently xapi only fetches a host's cpu info when it starts up, so this patch doesn't change the current behaviour (see commit message). Indeed a toolstack restart is needed to pick up a change in host features. This can be addressed in a separate PR.